spec/controllers/queries_controller_spec.rb

```diff
@@ -13,238 +11,362 @@ def response_json
-      SET TRANSACTION READ WRITE;
-      UPDATE posts SET cooked = '<p>you may already be a winner!</p>' WHERE id = #{p.id};
-      SAVEPOINT active_record_1;
-      SET TRANSACTION READ ONLY;
-      WITH query AS (
-        SELECT 1
-      SQL
-
-      run_query q.id
-      p.reload
-
-      # Manual Test - change out the following line:
-      #
-      #  module ::DataExplorer
-      #   def self.run_query(...)
-      #     if query.sql =~ /;/
-      #
-      # to
-      #
-      #     if false && query.sql =~ /;/
-      #
-      # Afterwards, this test should fail on the below check.
-      expect(p.cooked).to_not match(/winner/)
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/semicolon/)
-    end
-
-    it "doesn't allow you to lock rows" do
-      q = make_query <<~SQL
-      SELECT id FROM posts FOR UPDATE
-      SQL
-
-      run_query q.id
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/read-only transaction/)
-    end
-
-    it "doesn't allow you to create a table" do
-      q = make_query <<~SQL
-      CREATE TABLE mytable (id serial)
-      SQL
-
-      run_query q.id
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/read-only transaction|syntax error/)
-    end
-
-    it "doesn't allow you to break the transaction" do
-      q = make_query <<~SQL
-      COMMIT
-      SQL
-
-      run_query q.id
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/syntax error/)
-
-      q.sql = <<~SQL
-      )
-      SQL
-
-      run_query q.id
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/syntax error/)
-
-      q.sql = <<~SQL
-      RELEASE SAVEPOINT active_record_1
-      SQL
-
-      run_query q.id
-      expect(response.status).to eq(422)
-      expect(response_json['errors']).to_not eq([])
-      expect(response_json['success']).to eq(false)
-      expect(response_json['errors'].first).to match(/syntax error/)
-    end
-
-    it "can export data in CSV format" do
-      q = make_query('SELECT 23 as my_value')
-      post :run, params: { id: q.id, download: 1 }, format: :csv
-      expect(response.status).to eq(200)
+  describe "#user_can_access_query" do
+    let(:group) { Fabricate(:group) }
+
+    before do
+      controller
+        .stubs(:params)
```

Comment:
I much prefer no stubs here, can this be avoided?

Reply:
> Overall this looks great, I also love the UX and positioning! great job
> 
> General pattern though it is a bit hard for me to review every line of the spec file, but can you confirm the new public controller actions all have an integration test to ensure users who are not allowed have no access and one's allowed do.

@SamSaffron I have added integration tests (before I had filters, and tested those, but never actually hitting the endpoint that use the filters), and removed the stubbing.